src/libstd/sys/common/util.rs

```diff
@@ -46,11 +46,24 @@ unsafe fn abort_internal() -> ! {
     libc::abort()
 }
 
-// On Windows, we want to avoid using libc, and there isn't a direct
-// equivalent of libc::abort.  The __failfast intrinsic may be a reasonable
-// substitute, but desireability of using it over the abort instrinsic is
-// debateable; see https://github.com/rust-lang/rust/pull/31519 for details.
-#[cfg(not(unix))]
+// On windows, use the processor-specific __fastfail mechanism
+// https://msdn.microsoft.com/en-us/library/dn774154.aspx
+#[cfg(all(windows, target_arch = "x86"))]
+unsafe fn abort_internal() -> ! {
+    asm!("int $$0x29" :: "{ecx}"(7) ::: volatile); // 7 is FAST_FAIL_FATAL_APP_EXIT
+    ::intrinsics::unreachable();
+}
+
+#[cfg(all(windows, target_arch = "x86_64"))]
+unsafe fn abort_internal() -> ! {
+    asm!("int $$0x29" :: "{rcx}"(7) ::: volatile); // 7 is FAST_FAIL_FATAL_APP_EXIT
```

Comment:
Technically speaking, on x86_64 if you move 7 into ecx, it'll automatically zero extend it to rcx, so you can use a single ecx implementation for both x86 and x86_64.